tests/core/usage_statistics/test_usage_statistics_data_context.py

```diff
@@ -0,0 +1,125 @@
+"""
+TODO
+"""
+from __future__ import annotations
+
+import ast
+import logging
+from unittest import mock
+
+import pytest
+
+from great_expectations.core.usage_statistics.events import UsageStatsEvents
+from great_expectations.data_context.data_context.cloud_data_context import (
+    CloudDataContext,
+)
+from great_expectations.data_context.data_context.ephemeral_data_context import (
+    EphemeralDataContext,
+)
+from great_expectations.data_context.data_context.file_data_context import (
+    FileDataContext,
+)
+from great_expectations.data_context.util import file_relative_path
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture
+def enable_usage_stats(monkeypatch):
+    monkeypatch.delenv("GE_USAGE_STATS")
+
+
+@pytest.fixture
+def usage_stats_decorated_methods_on_abstract_data_context() -> list[str]:
+    """
+    Uses AST parsing to collect the names of all methods that are decorated with
+    `usage_statistics_enabled_method`.
+
+    This is used to ensure that any changes made to `AbstractDataContext` are automatically
+    accounted for in our tests.
+    """
+    path_to_abstract_context = file_relative_path(
+        __file__,
+        "../../../great_expectations/data_context/data_context/abstract_data_context.py",
+    )
+    with open(path_to_abstract_context) as f:
+        root = ast.parse(f.read())
+
+    relevant_methods = []
+    for node in ast.walk(root):
+        if not isinstance(node, ast.FunctionDef):
+            continue
+
+        name = node.name
+        # Ignore dunders like __init__
+        if name.startswith("__") and name.endswith("__"):
+            continue
+
+        decorators = node.decorator_list
+        for decorator in decorators:
+            if isinstance(decorator, ast.Call):
+                if decorator.func.id == "usage_statistics_enabled_method":
+                    relevant_methods.append(name)
+                    break
+
+    return relevant_methods
+
+
+@mock.patch(
+    "great_expectations.core.usage_statistics.usage_statistics.UsageStatisticsHandler.emit"
+)
+@pytest.mark.integration
+def test_all_relevant_context_methods_emit_usage_stats(
+    mock_emit: mock.MagicMock,
+    usage_stats_decorated_methods_on_abstract_data_context: list[str],
+    enable_usage_stats,  # Needs to be before context fixtures to ensure usage stats handlers are attached
+    in_memory_runtime_context: EphemeralDataContext,
+    empty_data_context: FileDataContext,
+    empty_data_context_in_cloud_mode: CloudDataContext,
```

Comment:
As it stands, this test takes 30.1 seconds - 30 seconds for the setup of these 3 fixtures and 0.1 seconds for the actual logic below.

I'm working to clean this up.